static_assert failed when building debug version with clang-cl

## Description

System: Windows 10 64 bit, cmake, clang-cl, debug build.

When cmake building the debug version of the library, it stops with error 'static_assert failed due to requirement 'std::is_integral<std::_Container_proxy>::value' "T must be integral type"'.

## Steps to Reproduce

1. Clone latest rnp main branch and update submodules.
2. cmake ..\rnp_release -D CMAKE_BUILD_TYPE=Debug -D CMAKE_INSTALL_PREFIX="INSTALL" -D BUILD_SHARED_LIBS=on -D BUILD_TESTING=off -D CRYPTO_BACKEND="openssl" -D ENABLE_COVERAGE=off -D DIRENT_INCLUDE_DIR="C:\Users\C081745\OneDrive\Desktop\rnp_release\include\dirent" -D GETOPT_INCLUDE_DIR="C:\Users\C081745\OneDrive\Desktop\rnp_release\include\getopt" -D GETOPT_LIBRARY="C:\Users\C081745\OneDrive\Desktop\rnp_release\include\getopt\getoptd.lib" -D CMAKE_PREFIX_PATH="C:\Users\C081745\OneDrive\Desktop\tplib64_clang_x64\openssl;C:\Users\C081745\OneDrive\Desktop\tplib64_clang_x64\zlib;C:\Users\C081745\OneDrive\Desktop\bzip2;C:\Users\C081745\OneDrive\Desktop\json-c_all\build_debug\INSTALL" -G "Visual Studio 16 2019" -T ClangCL -A x64
3. cmake --build . --config Debug

## Expected Behavior

To build successfully.

## Actual Behavior

Building stop with:
`C:\Users\C081745\OneDrive\Desktop\rnp_release\src\lib\crypto/mem.h(47,5): error : static_assert failed due to requirement 'std::is_integral<std::_Container_proxy>::value' "T must be integral type" [C:
\Users\C081745\OneDrive\Desktop\rnp_build_debug\src\lib\librnp-obj.vcxproj]`

And another error:
`C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.28.29910\include\vector(682,27): error : no matching conversion for static_cast from 'std::vector<unsigned char, rnp::ossl_alloc
ator<unsigned char>>::_Alty' (aka 'ossl_allocator<unsigned char>') to '_Rebind_alloc_t<std::vector<unsigned char, rnp::ossl_allocator<unsigned char>>::_Alty, std::_Container_proxy>' (aka 'ossl_allocator<std::_C
ontainer_proxy>') [C:\Users\C081745\OneDrive\Desktop\rnp_build_debug\src\lib\librnp-obj.vcxproj]`

[log.txt](https://github.com/rnpgp/rnp/files/11272088/log.txt)

Thank you.

I changed the compiler to msvc (still trying to build debug), but got same error.

`C:\Users\C081745\OneDrive\Desktop\rnp_release\src\lib\crypto\mem.h(47,40): error C2338: T must be integral type [C:\Users\C081745\OneDrive\Desktop\rnp_msvc_build_debug\src\lib\librnp-obj.vcxproj]`

--- a/.github/workflows/windows-native.yml
+++ b/.github/workflows/windows-native.yml
@@ -72,24 +72,14 @@ jobs:
         arch: [ { name: 'x64',    triplet: 'x64-windows' } ]
         toolset: [ 'v142',  'ClangCL' ]
         backend: [ 'botan', 'openssl' ]
-        shared_libs: [ 'on', 'off']
-        use_cmake_prefix_path: [ 'off' ]
+        shared_libs: [ 'off']
+        use_cmake_prefix_path: [ 'on', 'off' ]
         include:
-          - arch:  { name: 'x64', triplet: 'x64-windows' }
-            toolset:                'ClangCL'
-            backend:                'openssl'
-            use_cmake_prefix_path:  'on'
-            shared_libs:            'on'
-          - arch:  { name: 'x64', triplet: 'x64-windows' }
-            toolset:                'ClangCL'
-            backend:                'openssl'
-            use_cmake_prefix_path:  'on'
-            shared_libs:            'off'
           - arch:  { name: 'Win32',  triplet: 'x86-windows' }
             toolset:                'ClangCL'
             backend:                'botan'
-            use_cmake_prefix_path:  'off'
-            shared_libs:            'on'
+            use_cmake_prefix_path:  'on'
+            shared_libs:            'off'
           - arch:  { name: 'Win32',  triplet: 'x86-windows' }
             toolset:                'v142'
             backend:                'openssl'
@@ -142,7 +132,7 @@ jobs:
           cmake -B build   -G "Visual Studio 16 2019"  \
                            -A ${{ matrix.arch.name }}  \
                            -T ${{ matrix.toolset }}    \
-                           -DBUILD_SHARED_LIBS=${{ matrix.shared_lib}} \
+                           -DBUILD_SHARED_LIBS=${{ matrix.shared_libs}} \
                            -DCRYPTO_BACKEND=${{ matrix.backend }}      \
                            -DCMAKE_TOOLCHAIN_FILE=${{ env.VCPKG_DIR }}/scripts/buildsystems/vcpkg.cmake .
 
@@ -154,7 +144,7 @@ jobs:
           cmake -B build   -G "Visual Studio 16 2019"  \
                            -A ${{ matrix.arch.name }}  \
                            -T ${{ matrix.toolset }}    \
-                           -DBUILD_SHARED_LIBS=${{ matrix.shared_lib}} \
+                           -DBUILD_SHARED_LIBS=${{ matrix.shared_libs}} \
                            -DCRYPTO_BACKEND=${{ matrix.backend }}      \
                            -DCMAKE_PREFIX_PATH=${{ env.VCPKG_DIR }}/installed/${{ matrix.arch.triplet }} .
           echo ${{ env.VCPKG_DIR }}/installed/${{ matrix.arch.triplet }}/bin >> $GITHUB_PATH

--- a/ci/tests/ci-tests.sh
+++ b/ci/tests/ci-tests.sh
@@ -106,7 +106,7 @@ test_supported_features() {
         ;;
       *)
         so_folder="lib64"
-        botan_only+=("${brainpool[@]}")
+#       botan_only+=("${brainpool[@]}")
     esac
 
     if [[ "${CRYPTO_BACKEND:-}" == "openssl" ]]; then

--- a/src/lib/crypto/mem.h
+++ b/src/lib/crypto/mem.h
@@ -1,5 +1,5 @@
 /*-
- * Copyright (c) 2021 Ribose Inc.
+ * Copyright (c) 2021-2023 Ribose Inc.
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -44,7 +44,10 @@ template <typename T> using secure_vector = Botan::secure_vector<T>;
 #elif defined(CRYPTO_BACKEND_OPENSSL)
 template <typename T> class ossl_allocator {
   public:
-    static_assert(std::is_integral<T>::value, "T must be integral type");
+#if !defined(_MSC_VER) || !defined(_DEBUG)
+    /* MSVC in debug mode uses non-integral proxy types in container types */
+    static_assert(std::is_integral<T>::value, "secure_vector can hold integral types only");
+#endif
 
     typedef T           value_type;
     typedef std::size_t size_type;
@@ -99,7 +102,11 @@ template <typename T> using secure_vector = std::vector<T, ossl_allocator<T> >;
 
 template <typename T, std::size_t N> struct secure_array {
   private:
-    static_assert(std::is_integral<T>::value, "T must be integer type");
+#if !defined(_MSC_VER) || !defined(_DEBUG)
+    /* MSVC in debug mode uses non-integral proxy types in container types */
+    static_assert(std::is_integral<T>::value, "secure_array can hold integral types only");
+#endif
+
     std::array<T, N> data_;
 
   public:

